File: app/src/main/java/com/boardgamegeek/sorter/PlaysLengthSorter.java
package com.boardgamegeek.sorter;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Plays;

public class PlaysLengthSorter extends PlaysSorter {
	private static final String AND_MORE_SUFFIX = "+ ";
	@NonNull private final String noLength;
	@NonNull private final String hoursSuffix;
	@NonNull private final String minutesSuffix;

	public PlaysLengthSorter(@NonNull Context context) {
		super(context);
		orderByClause = getClause(Plays.LENGTH, true);
		noLength = context.getString(R.string.no_length);
		hoursSuffix = context.getString(R.string.hours_abbr);
		minutesSuffix = context.getString(R.string.minutes_abbr);
	}

	@StringRes
	@Override
	protected int getDescriptionId() {
		return R.string.menu_plays_sort_length;
	}

	@Override
	public int getType() {
		return PlaysSorterFactory.TYPE_PLAY_LENGTH;
	}

	@Override
	public String[] getColumns() {
		return new String[] { Plays.LENGTH };
	}

	@NonNull
	@Override
	public String getHeaderText(@NonNull Cursor cursor) {
		int minutes = getInt(cursor, Plays.LENGTH);
		if (minutes == 0) {
			return noLength;
		}
		if (minutes >= 120) {
			return (minutes / 60) + AND_MORE_SUFFIX + hoursSuffix;
		} else if (minutes >= 60) {
			return (minutes / 10) * 10 + AND_MORE_SUFFIX + minutesSuffix;
		} else if (minutes >= 30) {
			return (minutes / 5) * 5 + AND_MORE_SUFFIX + minutesSuffix;
		} else {
			return minutes + minutesSuffix;
		}
	}
}

File: app/src/main/java/com/boardgamegeek/sorter/PlaysLengthSorter.kt
package com.boardgamegeek.sorter

import android.content.Context
import androidx.annotation.StringRes
import com.boardgamegeek.R
import com.boardgamegeek.provider.BggContract.Plays
import com.boardgamegeek.ui.model.PlayModel
import java.text.NumberFormat

class PlaysLengthSorter(context: Context) : PlaysSorter(context) {
    private val numberFormat = NumberFormat.getNumberInstance()
    private val noLength = context.getString(R.string.no_length)
    private val hoursSuffix = context.getString(R.string.hours_abbr)
    private val minutesSuffix = context.getString(R.string.minutes_abbr)

    override val descriptionResId: Int
        @StringRes
        get() = R.string.menu_plays_sort_length

    override val type: Int
        get() = PlaysSorterFactory.TYPE_PLAY_LENGTH

    override val sortColumn: String
        get() = Plays.LENGTH

    override val isSortDescending: Boolean
        get() = true

    override fun getSectionText(play: PlayModel): String {
        val minutes = play.length
        return when {
            minutes == 0 -> noLength
            minutes >= 120 -> numberFormat.format(minutes / 60) + AND_MORE_SUFFIX + hoursSuffix
            minutes >= 60 -> numberFormat.format(minutes / 10 * 10) + AND_MORE_SUFFIX + minutesSuffix
            minutes >= 30 -> numberFormat.format(minutes / 5 * 5) + AND_MORE_SUFFIX + minutesSuffix
            else -> numberFormat.format(minutes) + minutesSuffix
        }
    }

    companion object {
        private const val AND_MORE_SUFFIX = "+ "
    }
}
